### PR TITLE
MULE-18794: Cannot access application resources from KeyResolver for Configurations Part 2

### PR DESCRIPTION
**Pull request by SebaElizalde**

### CODE REVIEW

**Comment by SebaElizalde** [comment]
--validate

**Comment by SebaElizalde** [comment]
--validate

File: tests/test-extensions/classloading-extension/src/main/java/org/mule/test/classloading/ClassLoadingExtension.java [hunk 1]
```diff
@@ -11,12 +11,14 @@
...
 import org.mule.runtime.extension.api.annotation.connectivity.ConnectionProviders;
 import org.mule.runtime.extension.api.annotation.dsl.xml.Xml;
+import org.mule.runtime.extension.api.annotation.metadata.MetadataScope;
 import org.mule.test.classloading.api.ClassLoadingHelper;
 
 @Extension(name = "ClassLoading")
 @Xml(prefix = "classloading")
 @ConnectionProviders({CLPoolingConnectionProvider.class, CLCachedConnectionProvider.class, CLNoneConnectionProvider.class})
 @Configurations(CLConfiguration.class)
+@MetadataScope(keysResolver = CLKeysResolver.class)
```
Comment (ndinu) [inline tests/test-extensions/classloading-extension/src/main/java/org/mule/test/classloading/ClassLoadingExtension.java:21]: I would set the type resolver for a specific operation you want to modify, or create a new one.

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/classloading/ClassLoadingOnConnectionsTestCase.java [hunk 2]
```diff
@@ -29,13 +34,18 @@
...
+
 import org.hamcrest.Matcher;
 import org.hamcrest.core.StringContains;
 import org.junit.Rule;
 import org.junit.Test;
 
 public class ClassLoadingOnConnectionsTestCase extends AbstractExtensionFunctionalTestCase {
 
+  @Inject
+  private MetadataService metadataManager;
```
Comment (ndinu) [inline modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/classloading/ClassLoadingOnConnectionsTestCase.java:47]: call the field metadataService

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/classloading/ClassLoadingOnConnectionsTestCase.java [hunk 3]
```diff
@@ -72,6 +82,12 @@ public void allOptionalParameterGroup() throws Exception {
     verifyUsedClassLoaders(ALL_OPTIONAL_PARAMETER_GROUP);
   }
 
+  @Test
+  public void operationWithMetadataResolver() throws Exception {
```
Comment (ndinu) [inline modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/classloading/ClassLoadingOnConnectionsTestCase.java:86]: I don't feel that this test belongs in this class.

File: tests/test-extensions/classloading-extension/src/main/java/org/mule/test/classloading/CLKeysResolver.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import org.mule.runtime.api.metadata.MetadataKeyBuilder;
+import org.mule.runtime.api.metadata.MetadataResolvingException;
+import org.mule.runtime.api.metadata.resolving.OutputTypeResolver;
+import org.mule.runtime.api.metadata.resolving.TypeKeysResolver;
+
+import java.util.Set;
+
+import com.google.common.collect.ImmutableSet;
+
+public class CLKeysResolver implements TypeKeysResolver, OutputTypeResolver {
```
Comment (ndinu) [inline tests/test-extensions/classloading-extension/src/main/java/org/mule/test/classloading/CLKeysResolver.java:26]: If you are going to assert that the operation metadata key resolution and source metadata key resolution are using the correct class loader, I would make them use two different resolvers, and use the addClassLoader method with different values.

**Review by ndinu — CHANGES_REQUESTED** [review]

File: tests/test-extensions/classloading-extension/src/main/java/org/mule/test/classloading/CLKeysResolver.java (hunk 4 above)
Comment (SebaElizalde) [reply tests/test-extensions/classloading-extension/src/main/java/org/mule/test/classloading/CLKeysResolver.java:26]: They both end up calling the same method, so I don't feel like there is a need to distinguish between sources and operations by making 2 separate tests.

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/classloading/ClassLoadingOnConnectionsTestCase.java (hunk 3 above)
Comment (SebaElizalde) [reply modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/classloading/ClassLoadingOnConnectionsTestCase.java:86]: will refactor

File: tests/test-extensions/classloading-extension/src/main/java/org/mule/test/classloading/ClassLoadingExtension.java (hunk 1 above)
Comment (SebaElizalde) [reply tests/test-extensions/classloading-extension/src/main/java/org/mule/test/classloading/ClassLoadingExtension.java:21]: This is needed for configs to be annotated with the KeysResolver annotation in the model. That then allows the following line to be called in the tests:
`metadataService.getMetadataKeys(builder().globalName("none").build());`
and that will trigger getting the metadata keys.

File: tests/test-extensions/classloading-extension/src/main/java/org/mule/test/classloading/CLKeysResolver.java (hunk 4 above)
Comment (ndinu) [reply tests/test-extensions/classloading-extension/src/main/java/org/mule/test/classloading/CLKeysResolver.java:26]: I think that calling the same method is an implementation detail, we should cover both cases.

**Comment by SebaElizalde** [comment]
--validate

**Review by ndinu — DISMISSED** [review]

**Review by ndinu — APPROVED** [review]

**Pull request merged**